web/src/dropdown_widget.js

```diff
@@ -1,176 +1,271 @@
-    // Used to focus the `target` after dropdown is closed. This is important since the dropdown is
-    // appended to `body` and hence `body` is focused when the dropdown is closed, which makes
-    // it hard for the user to get focus back to the `target`.
-    const focus_target_on_hidden = dropdown_props.focus_target_on_hidden || true;
-    // Should enter keypress on target show the dropdown.
-    const show_on_target_enter_keypress = dropdown_props.show_on_target_enter_keypress || false;
-
-    if (show_on_target_enter_keypress) {
-        $("body").on("keypress", tippy_props.target, (e) => {
-            if (e.key === "Enter") {
-                $(tippy_props.target).trigger("click");
-                e.stopPropagation();
-                e.preventDefault();
-            }
-        });
+export class DropdownWidget {
+    constructor({
+        widget_name,
+        get_options,
+        item_click_callback,
+        // Provide an parent element to widget which will be re-rendered if the widget is setup again.
+        // It is important to not pass `$("body")` here for widgets that would be `setup()`
+        // multiple times, so that we don't have duplicate event handlers.
+        $events_container,
+        on_show_callback = noop,
+        on_mount_callback = noop,
+        on_hidden_callback = noop,
+        on_exit_with_escape_callback = noop,
+        render_selected_option = noop,
+        // Used to focus the `target` after dropdown is closed. This is important since the dropdown is
+        // appended to `body` and hence `body` is focused when the dropdown is closed, which makes
+        // it hard for the user to get focus back to the `target`.
+        focus_target_on_hidden = true,
+        tippy_props = {},
+        // NOTE: Any value other than `null` will be rendered when class is initialized.
+        default_id = null,
+        unique_id_type = null,
+    }) {
+        this.widget_name = widget_name;
+        this.widget_id = `#${CSS.escape(widget_name)}_widget`;
+        // A widget wrapper may not exist based on the UI requirement.
+        this.widget_wrapper_id = `${this.widget_id}_wrapper`;
+        this.widget_value_selector = `${this.widget_id} .dropdown_widget_value`;
+        this.get_options = get_options;
+        this.item_click_callback = item_click_callback;
+        this.focus_target_on_hidden = focus_target_on_hidden;
+        this.on_show_callback = on_show_callback;
+        this.on_mount_callback = on_mount_callback;
+        this.on_hidden_callback = on_hidden_callback;
+        this.on_exit_with_escape_callback = on_exit_with_escape_callback;
+        this.render_selected_option = render_selected_option;
+        this.tippy_props = tippy_props;
+        this.list_widget = null;
+        this.instance = null;
+        this.default_id = default_id;
+        this.current_value = default_id;
+        this.unique_id_type = unique_id_type;
+        this.$events_container = $events_container;
     }
```

Comment:
I don't know whether you accidentally squashed something, but this commit seems to do most of the migrations in addition to switching to using `Class`.

Reply:
Yes, that is intentional since I wanted to know what class desing would work well for all the dropdown cases that we have. So, I migrated all of them first and then split them into commits.